src/high-level/abstract-tensor.lisp

```diff
@@ -0,0 +1,269 @@
+  (:documentation "Set the value of element at position"))
+
+(defgeneric copy-tensor (tensor &rest args)
+  (:documentation "Create a new tensor with the same properties as the given tensor, not initializing the storage"))
+
+(defgeneric deep-copy-tensor (tensor &rest args)
+  (:documentation "Create a new tensor with the same properties as the given tensor, copying storage"))
+
+;;; abstract-tensor generic methods
+;;; Can be optimized by subclasses
+
+(defgeneric rank (tensor)
+  (:documentation "Rank (number of dimensions) of the tensor")
+  (:method ((tensor abstract-tensor))
+    (length (shape tensor))))
+
+(defgeneric size (tensor)
+  (:documentation "Total number of elements in the tensor")
+  (:method ((tensor abstract-tensor))
+    (reduce #'* (shape tensor))))
+
+(defgeneric element-type (tensor)
+  (:documentation "The type of the elements in the tensor")
+  (:method ((tensor abstract-tensor))
+    (declare (ignore tensor))
+    t))
+
+(defgeneric lisp-array (tensor &optional target)
+  (:documentation "Return a lisp array containing the data from the tensor
+
+If TARGET is specified then the contents of the tensor are copied into the array.
+In the event TARGET is not specified, the result may return an array sharing memory with the input tensor.")
+  (:method :before (tensor &optional target)
+    (unless (null target)
+      (policy-cond:policy-if
+       (< speed safety)
+       (assert (and (= (rank tensor) (array-rank target))
+                    (equal (shape tensor) (array-dimensions target))))
+       nil)))
+  (:method ((tensor abstract-tensor) &optional target)
+    (let  ((arr (or target (make-array (shape tensor) :element-type (element-type tensor)))))
+      (map-indexes
+       (shape tensor)
+       (lambda (&rest pos)
+         (let ((val (apply #'tref tensor pos)))
+           (apply #'(setf aref) val arr pos))))
+      arr)))
+
+(defgeneric map! (function tensor)
+  (:documentation "Map elements of TENSOR by replacing the value with the output of FUNCTION on the element")
+  (:method ((function function) (tensor abstract-tensor))
+    (map-indexes
+     (shape tensor)
+     (lambda (&rest dims)
+       (apply #'(setf tref) (funcall function (apply #'tref tensor dims)) tensor dims)))
+    tensor))
+
+(defgeneric into! (function tensor &key order)
+  (:documentation "Map indices of TENSOR by replacing the value with the output of FUNCTION on the index at the element.
+
+If ORDER is specified then traverse TENSOR in the specified order (column major or row major).")
+  (:method ((function function) (tensor abstract-tensor) &key (order :row-major))
+    (let ((map-func (if (eql order :column-major)
+                        #'map-column-indexes
+                        #'map-indexes)))
+      (funcall map-func
+               (shape tensor)
+               (lambda (&rest dims)
+                 (apply #'(setf tref) (apply function dims) tensor dims))))
+    tensor))
+
+(defgeneric foreach (function tensor &key order)
```

Comment:
I don't think `order` should be selectable for `foreach`. Is there a reason?

Reply:
It seems as though `into!` and `foreach` should both drop that